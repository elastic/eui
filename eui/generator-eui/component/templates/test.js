import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test';

import { <%= componentName %> } from './<%= fileName %>';

describe('<%= componentName %>', () => {
  test('is rendered', () => {
    const component = render(
      <<%= componentName %> {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
