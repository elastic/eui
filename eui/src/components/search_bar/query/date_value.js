import { isDateLike, isNumber } from '../../../services/predicate';
import {
  dateFormat as defaultDateFormat,
  dateGranularity
} from './date_format';
import moment from 'moment';

export const DATE_TYPE = 'date';

export const dateValuesEqual = (v1, v2) => {
  return v1.raw === v2.raw &&
    v1.granularity === v2.granularity &&
    v1.text === v2.text;
};

export const isDateValue = (value) => {
  return !!value && value.type === DATE_TYPE && !!value.raw && !!value.text && !!value.resolve;
};

export const dateValue = (raw, granularity, dateFormat = defaultDateFormat) => {
  if (!raw) {
    return undefined;
  }
  if (isDateLike(raw)) {
    return {
      type: DATE_TYPE,
      raw,
      granularity,
      text: dateFormat.print(raw),
      resolve: () => moment(raw)
    };
  }
  if (isNumber(raw)) {
    return {
      type: DATE_TYPE,
      raw,
      granularity,
      text: raw.toString(),
      resolve: () => moment(raw)
    };
  }
  const text = raw.toString();
  return {
    type: DATE_TYPE,
    raw,
    granularity,
    text,
    resolve: () => dateFormat.parse(text)
  };
};

export const dateValueParser = (format = defaultDateFormat) => {
  return (text) => {
    const parsed = format.parse(text);
    return dateValue(text, dateGranularity(parsed), format);
  };
};
