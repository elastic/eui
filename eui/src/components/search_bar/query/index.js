export { Query } from './query';
export { AST } from './ast';
export {
  dateValueParser as parseDateValue
} from './date_value';
