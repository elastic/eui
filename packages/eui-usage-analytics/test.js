/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const Codeowners = require('codeowners');
const temp = new Codeowners("../../../kibana");

const { scan } = require('./scan');

const runScan = async () => {
  const scanResult = await scan();
  console.log(scanResult);
};

runScan();
