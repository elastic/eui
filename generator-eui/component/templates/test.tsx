/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from '../../test/rtl';
import { requiredProps } from '../../test/required_props';

import { <%= componentName %> } from './<%= fileName %>';

describe('<%= componentName %>', () => {
  it('renders', () => {
    const { container } = render(
      <<%= componentName %> {...requiredProps} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
