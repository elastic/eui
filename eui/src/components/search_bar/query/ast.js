import { isArray, isNil } from '../../../services/predicate';
import { isDateValue, dateValuesEqual } from './date_value';

export const Match = Object.freeze({
  MUST: 'must',
  MUST_NOT: 'must_not',
  isMust(match) {
    return match === Match.MUST;
  },
  isMustClause(clause) {
    return Match.isMust(clause.match);
  }
});

export const Operator = Object.freeze({
  EQ: 'eq',
  GT: 'gt',
  GTE: 'gte',
  LT: 'lt',
  LTE: 'lte',
  isEQ(match) {
    return match === Operator.EQ;
  },
  isEQClause(clause) {
    return Operator.isEQ(clause.operator);
  },
  isRange(match) {
    return Operator.isGT(match) || Operator.isGTE(match) || Operator.isLT(match) || Operator.isLTE(match);
  },
  isRangeClause(clause) {
    return Operator.isRange(clause.operator);
  },
  isGT(match) {
    return match === Operator.GT;
  },
  isGTClause(clause) {
    return Operator.isGT(clause.operator);
  },
  isGTE(match) {
    return match === Operator.GTE;
  },
  isGTEClause(clause) {
    return Operator.isGTE(clause.operator);
  },
  isLT(match) {
    return match === Operator.LT;
  },
  isLTClause(clause) {
    return Operator.isLT(clause.operator);
  },
  isLTE(match) {
    return match === Operator.LTE;
  },
  isLTEClause(clause) {
    return Operator.isLTE(clause.operator);
  }
});

const Term = Object.freeze({
  TYPE: 'term',
  isInstance: (clause) => {
    return clause.type === Term.TYPE;
  },
  must: (value) => {
    return { type: Term.TYPE, value, match: Match.MUST };
  },
  mustNot: (value) => {
    return { type: Term.TYPE, value, match: Match.MUST_NOT };
  }
});

const Field = Object.freeze({
  TYPE: 'field',
  isInstance: (clause) => {
    return clause.type === Field.TYPE;
  },
  must: {
    eq: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST, operator: Operator.EQ }),
    gt: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST, operator: Operator.GT }),
    gte: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST, operator: Operator.GTE }),
    lt: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST, operator: Operator.LT }),
    lte: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST, operator: Operator.LTE })
  },
  mustNot: {
    eq: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST_NOT, operator: Operator.EQ }),
    gt: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST_NOT, operator: Operator.GT }),
    gte: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST_NOT, operator: Operator.GTE }),
    lt: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST_NOT, operator: Operator.LT }),
    lte: (field, value) => ({ type: Field.TYPE, field, value, match: Match.MUST_NOT, operator: Operator.LTE })
  }
});

const Is = Object.freeze({
  TYPE: 'is',
  isInstance: (clause) => {
    return clause.type === Is.TYPE;
  },
  must: (flag) => {
    return { type: Is.TYPE, flag, match: Match.MUST };
  },
  mustNot: (flag) => {
    return { type: Is.TYPE, flag, match: Match.MUST_NOT };
  }
});

const valuesEqual = (v1, v2) => {
  if (isDateValue(v1)) {
    return dateValuesEqual(v1, v2);
  }
  return v1 === v2;
};

const arrayIncludesValue = (array, value) => {
  return array.some(item => valuesEqual(item, value));
};

/**
 * The AST structure is an array of clauses. There are 3 types of clauses that are supported:
 *
 * :term:
 * Holds a VALUE and an OCCUR. The OCCUR indicates whether the value must match or must not match. Default
 * clauses are not associated with any specific field - when executing the search, one can specify what are
 * the default fields that the default clauses will be matched against.
 *
 * :field:
 * Like the `term` clause, holds a VALUE and an MATCH, but this clause also specifies the field that the
 * value will be matched against.
 *
 * :is:
 * Holds a FLAG and indicates whether this flag must be applied or must not be applied. Typically this clause
 * matches against boolean values of a record (e.g. "is:online", "is:internal", "is:on", etc..)
 *
 * This AST is immutable - every "mutating" operation returns a newly mutated AST.
 */
export class _AST {

  static create(clauses) {
    return new _AST(clauses);
  }

  constructor(clauses = []) {
    this._clauses = clauses;
    this._indexedClauses = clauses.reduce((map, clause) => {
      switch (clause.type) {
        case Field.TYPE:
          if (!map.field[clause.field]) {
            map.field[clause.field] = [];
          }
          map.field[clause.field].push(clause);
          return map;
        case Is.TYPE:
          map.is[clause.flag] = clause;
          return map;
        case Term.TYPE:
          map.term.push(clause);
          return map;
        default:
          throw new Error(`Unknown query clause type [${clause.type}]`);
      }
    }, { field: {}, is: {}, term: [] });
  }

  get clauses() {
    return this._clauses;
  }

  getTermClauses() {
    return this._indexedClauses.term;
  }

  getTermClause(value) {
    const clauses = this.getTermClauses();
    return clauses.find(clause => valuesEqual(clause.value, value));
  }

  getFieldNames() {
    return Object.keys(this._indexedClauses.field);
  }

  getFieldClauses(field = undefined) {
    return field ?
      this._indexedClauses.field[field] :
      this._clauses.filter(Field.isInstance);
  }

  getFieldClause(field, predicate) {
    const clauses = this.getFieldClauses(field);
    if (clauses) {
      return clauses.find(predicate);
    }
  }

  hasOrFieldClause(field, value = undefined) {
    const clauses = this.getFieldClause(field, clause => isArray(clause.value));
    if (!clauses) {
      return false;
    }
    return isNil(value) || clauses.some(clause => arrayIncludesValue(clause.value, value));
  }

  getOrFieldClause(field, value = undefined) {
    return this.getFieldClause(field, clause => isArray(clause.value) && (isNil(value) || arrayIncludesValue(clause.value, value)));
  }

  addOrFieldValue(field, value, must = true, operator = Operator.EQ) {
    const existingClause = this.getOrFieldClause(field);
    if (!existingClause) {
      const newClause = must ? Field.must[operator](field, [value]) : Field.mustNot[operator](field, [value]);
      return new _AST([...this._clauses, newClause]);
    }
    const clauses = this._clauses.map(clause => {
      if (clause === existingClause) {
        clause.value.push(value);
      }
      return clause;
    });
    return new _AST(clauses);
  }

  removeOrFieldValue(field, value) {
    const existingClause = this.getOrFieldClause(field, value);
    if (!existingClause) {
      return new _AST([...this._clauses]);
    }
    const clauses = this._clauses.reduce((clauses, clause) => {
      if (clause !== existingClause) {
        clauses.push(clause);
        return clauses;
      }
      const filteredValue = clause.value.filter(val => !valuesEqual(val, value));
      if (filteredValue.length === 0) {
        return clauses;
      }
      clauses.push({ ...clause, value: filteredValue });
      return clauses;
    }, []);
    return new _AST(clauses);
  }

  removeOrFieldClauses(field) {
    const clauses = this._clauses.filter(clause => {
      return !Field.isInstance(clause) || clause.field !== field || !isArray(clause.value);
    });
    return new _AST(clauses);
  }

  hasSimpleFieldClause(field, value = undefined) {
    const clauses = this.getFieldClause(field, clause => !isArray(clause.value));
    if (!clauses) {
      return false;
    }
    return isNil(value) || clauses.some(clause => valuesEqual(clause.value, value));
  }

  getSimpleFieldClause(field, value = undefined) {
    return this.getFieldClause(field, clause => !isArray(clause.value) && (isNil(value) || valuesEqual(clause.value, value)));
  }

  addSimpleFieldValue(field, value, must = true, operator = Operator.EQ) {
    const clause = must ? Field.must[operator](field, value) : Field.mustNot[operator](field, value);
    return this.addClause(clause);
  }

  removeSimpleFieldValue(field, value) {
    const existingClause = this.getSimpleFieldClause(field, value);
    if (!existingClause) {
      return new _AST([...this._clauses]);
    }
    const clauses = this._clauses.filter(clause => clause !== existingClause);
    return new _AST(clauses);
  }

  removeSimpleFieldClauses(field) {
    const clauses = this._clauses.filter(clause => {
      return !Field.isInstance(clause) || clause.field !== field || isArray(clause.value);
    });
    return new _AST(clauses);
  }

  getIsClauses() {
    return Object.values(this._indexedClauses.is);
  }

  getIsClause(flag) {
    return this._indexedClauses.is[flag];
  }

  removeIsClause(flag) {
    return new _AST(this._clauses.filter(clause => !Is.isInstance(clause) || clause.flag !== flag));
  }

  /**
   * Creates and returns a new AST with the given clause added to the current clauses. If
   * the current clauses already include a similar clause, it will be (in-place) replaced by
   * the given clause. Whether a clause is similar to the given one depends on the type of the clause.
   * Two clauses are similar if:
   *
   * - they are both of the same type
   * - if they are `default` clauses, they must have the same value
   * - if they are `term` clauses, they must have the same fields and values
   * - if they are `is` clauses, they must have the same flags
   *
   * The reasoning behind not including the `match` attributes of the clauses in the rules above, stems
   * in the fact that the AST clauses are ANDed, and having two similar clauses with two different
   * match attributes creates a logically contradicted AST (e.g. what does it mean to
   * "(must have x) AND (must not have x)"?)
   *
   * note:  in-place replacement means the given clause will be placed in the same position as the one it
   *        replaced
   */
  addClause(newClause) {
    let added = false;
    const newClauses = this._clauses.reduce((clauses, clause) => {
      if (newClause.type !== clause.type) {
        clauses.push(clause);
        return clauses;
      }
      switch (newClause.type) {
        case Term.TYPE:
          if (newClause.value !== clause.value) {
            clauses.push(clause);
            return clauses;
          }
          break;
        case Field.TYPE:
          if (newClause.field !== clause.field || newClause.value !== clause.value) {
            clauses.push(clause);
            return clauses;
          }
          break;
        case Is.TYPE:
          if (newClause.flag !== clause.flag) {
            clauses.push(clause);
            return clauses;
          }
          break;
        default:
          throw new Error(`unknown clause type [${newClause.type}]`);
      }
      added = true;
      clauses.push(newClause);
      return clauses;
    }, []);
    if (!added) {
      newClauses.push(newClause);
    }
    return new _AST(newClauses);
  }
}

export const AST = Object.freeze({
  Match,
  Operator,
  Term,
  Field,
  Is,
  create: (clauses) => new _AST(clauses)
});
