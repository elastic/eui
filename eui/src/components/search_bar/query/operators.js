import { dateFormat, dateGranularity } from './date_format';
import { isDateValue } from './date_value';
import {
  isArray, isBoolean,
  isNumber, isString,
  isDateLike, isNil
} from '../../../services/predicate';
import moment from 'moment';
const utc = moment.utc;

const resolveValueAsDate = (value) => {
  if (moment.isMoment(value)) {
    return value;
  }
  if (moment.isDate(value) || isNumber(value)) {
    return moment(value);
  }
  return dateFormat.parse(value.toString());
};

const defaultEqOptions = {
  ignoreCase: true
};

export const eq = (fieldValue, clauseValue, options = {}) => {
  options = { ...defaultEqOptions, ...options };

  if (isNil(fieldValue) || isNil(clauseValue)) {
    return fieldValue === clauseValue;
  }

  if (isDateValue(clauseValue)) {
    const dateFieldValue = resolveValueAsDate(fieldValue);
    if (clauseValue.granularity) {
      return clauseValue.granularity.isSame(dateFieldValue, clauseValue.resolve());
    }
    return dateFieldValue.isSame(clauseValue.resolve());
  }

  if (isString(fieldValue)) {
    return options.ignoreCase ?
      fieldValue.toLowerCase().includes(clauseValue.toString().toLowerCase()) :
      fieldValue.includes(clauseValue.toString());
  }

  if (isNumber(fieldValue)) {
    clauseValue = Number(clauseValue);
    return fieldValue === clauseValue;
  }

  if (isBoolean(fieldValue)) {
    return clauseValue === fieldValue;
  }

  if (isDateLike(fieldValue)) {
    const date = resolveValueAsDate(clauseValue);
    if (!date.isValid()) {
      return false;
    }
    const granularity = dateGranularity(date);
    if (!granularity) {
      return utc(fieldValue).isSame(date);
    }
    return granularity.isSame(fieldValue, date);
  }

  if (isArray(fieldValue)) {
    return fieldValue.some(item => eq(item, clauseValue, options));
  }

  return false; // unknown value type
};

const greaterThen = (fieldValue, clauseValue, inclusive = false) => {
  if (isDateValue(clauseValue)) {
    const clauseDateValue = clauseValue.resolve();
    if (!clauseValue.granularity) {
      return inclusive ? utc(fieldValue).isSameOrAfter(clauseDateValue) : utc(fieldValue).isAfter(clauseDateValue);
    }
    if (inclusive) {
      return utc(fieldValue).isSameOrAfter(clauseValue.granularity.start(clauseDateValue));
    }
    return utc(fieldValue).isSameOrAfter(clauseValue.granularity.startOfNext(clauseDateValue));
  }

  if (isString(fieldValue)) {
    const str = clauseValue.toString();
    return inclusive ? fieldValue >= str : fieldValue > str;
  }

  if (isNumber(fieldValue)) {
    const number = Number(clauseValue);
    return inclusive ? fieldValue >= number : fieldValue > number;
  }

  if (isDateLike(fieldValue)) {
    const date = resolveValueAsDate(clauseValue);
    const granularity = dateGranularity(date);
    if (!granularity) {
      return inclusive ? utc(fieldValue).isSameOrAfter(date) : utc(fieldValue).isAfter(date);
    }
    if (inclusive) {
      return utc(fieldValue).isSameOrAfter(granularity.start(date));
    }
    return utc(fieldValue).isSameOrAfter(granularity.startOfNext(date));
  }

  if (isArray(fieldValue)) {
    return fieldValue.all(item => greaterThen(item, clauseValue, inclusive));
  }

  return false; // unsupported value type
};

export const gt = (fieldValue, clauseValue) => {
  if (isNil(fieldValue) || isNil(clauseValue)) {
    return false;
  }
  return greaterThen(fieldValue, clauseValue);
};

export const gte = (fieldValue, clauseValue) => {
  if (isNil(fieldValue) || isNil(clauseValue)) {
    return fieldValue === clauseValue;
  }
  return greaterThen(fieldValue, clauseValue, true);
};

export const lt = (fieldValue, clauseValue) => {
  if (isNil(fieldValue) || isNil(clauseValue)) {
    return false;
  }
  return !greaterThen(fieldValue, clauseValue, true);
};

export const lte = (fieldValue, clauseValue) => {
  if (isNil(fieldValue) || isNil(clauseValue)) {
    return fieldValue === clauseValue;
  }
  return !greaterThen(fieldValue, clauseValue);
};
