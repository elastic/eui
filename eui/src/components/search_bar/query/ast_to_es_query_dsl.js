import { printIso8601 } from './date_format';
import { isDateValue, dateValue } from './date_value';
import { AST } from './ast';
import { isArray, isDateLike, isString } from '../../../services/predicate';

const processDateOperation = (value, operator) => {
  const { granularity, resolve } = value;
  let expression = printIso8601(resolve());
  if (!granularity) {
    return { operator, expression };
  }
  switch (operator) {
    case AST.Operator.GT:
      expression = `${expression}||+1${granularity.es}/${granularity.es}`;
      return { operator: AST.Operator.GTE, expression };
    case AST.Operator.GTE:
      expression = `${expression}||/${granularity.es}`;
      return { operator, expression };
    case AST.Operator.LT:
      expression = `${expression}||/${granularity.es}`;
      return { operator, expression };
    case AST.Operator.LTE:
      expression = `${expression}||+1${granularity.es}/${granularity.es}`;
      return { operator: AST.Operator.LT, expression };
    default:
      expression = `${expression}||/${granularity.es}`;
      return { expression };
  }
};

export const _termValuesToQuery = (values, options) => {
  const body = {
    query: values.join(' ')
  };
  if (body.query === '') {
    return;
  }
  if (options.defaultFields) {
    body.fields = options.defaultFields;
  }
  return {
    'simple_query_string': body
  };
};

export const _fieldValuesToQuery = (field, operations, andOr) => {
  const queries = [];

  Object.keys(operations).forEach(operator => {
    const values = operations[operator];
    switch (operator) {

      case AST.Operator.EQ:
        const { terms, phrases, dates } = values.reduce((tokenTypes, value) => {
          if (isDateValue(value)) {
            tokenTypes.dates.push(value);
          } else if (isDateLike(value)) {
            tokenTypes.dates.push(dateValue(value));
          } else if (isString(value) && value.match(/\s/)) {
            tokenTypes.phrases.push(value);
          } else {
            tokenTypes.terms.push(value);
          }
          return tokenTypes;
        }, { terms: [], phrases: [], dates: [] });

        if (terms.length > 0) {
          queries.push({
            match: {
              [field]: {
                query: terms.join(' '),
                operator: andOr
              }
            }
          });
        }

        if (phrases.length > 0) {
          queries.push(...phrases.map(phrase => ({
            match_phrase: {
              [field]: phrase
            }
          })));
        }

        if (dates.length > 0) {
          queries.push(...dates.map(value => ({
            match: {
              [field]: processDateOperation(value).expression
            }
          })));
        }

        break;

      default:

        values.forEach(value => {
          if (isDateValue(value)) {
            const operation = processDateOperation(value, operator);
            queries.push({
              range: {
                [field]: {
                  [operation.operator]: operation.expression
                }
              }
            });
          } else {
            queries.push({
              range: {
                [field]: {
                  [operator]: value
                }
              }
            });
          }
        });
    }
  });

  if (queries.length === 1) {
    return queries[0];
  }

  const key = andOr === 'and' ? 'must' : 'should';
  return {
    bool: {
      [key]: [...queries]
    }
  };
};

export const _isFlagToQuery = (flag, on) => {
  return {
    term: { [flag]: on }
  };
};

const collectTerms = (ast) => {
  return ast.getTermClauses().reduce((values, clause) => {
    if (AST.Match.isMustClause(clause)) {
      values.must.push(clause.value);
    } else {
      values.mustNot.push(clause.value);
    }
    return values;
  }, { must: [], mustNot: [] });
};

const collectFields = (ast) => {

  const fieldArray = (obj, field, operator) => {
    if (!obj[field]) {
      obj[field] = {};
    }
    if (!obj[field][operator]) {
      obj[field][operator] = [];
    }
    return obj[field][operator];
  };

  return ast.getFieldClauses().reduce((fields, clause) => {
    if (AST.Match.isMustClause(clause)) {
      if (isArray(clause.value)) {
        fieldArray(fields.must.or, clause.field, clause.operator).push(...clause.value);
      } else {
        fieldArray(fields.must.and, clause.field, clause.operator).push(clause.value);
      }
    } else {
      if (isArray(clause.value)) {
        fieldArray(fields.mustNot.or, clause.field, clause.operator).push(...clause.value);
      } else {
        fieldArray(fields.mustNot.and, clause.field, clause.operator).push(clause.value);
      }
    }
    return fields;
  }, {
    must: { and: {}, or: {}, },
    mustNot: { and: {}, or: {} }
  });
};

export const astToEsQueryDsl = (ast, options = {}) => {

  if (ast.clauses.length === 0) {
    return { match_all: {} };
  }

  const extraMustQueries = options.extraMustQueries || [];
  const extraMustNotQueries = options.extraMustNotQueries || [];
  const termValuesToQuery = options.termValuesToQuery || _termValuesToQuery;
  const fieldValuesToQuery = options.fieldValuesToQuery || _fieldValuesToQuery;
  const isFlagToQuery = options.isFlagToQuery || _isFlagToQuery;

  const terms = collectTerms(ast);
  const fields = collectFields(ast);

  const must = [];
  must.push(...extraMustQueries);
  const termMustQuery = termValuesToQuery(terms.must, options);
  if (termMustQuery) {
    must.push(termMustQuery);
  }
  Object.keys(fields.must.and).forEach(field => {
    must.push(fieldValuesToQuery(field, fields.must.and[field], 'and'));
  });
  Object.keys(fields.must.or).forEach(field => {
    must.push(fieldValuesToQuery(field, fields.must.or[field], 'or'));
  });
  ast.getIsClauses().forEach(clause => {
    must.push(isFlagToQuery(clause.flag, AST.Match.isMustClause(clause)));
  });

  const mustNot = [];
  mustNot.push(...extraMustNotQueries);
  const termMustNotQuery = termValuesToQuery(terms.mustNot, options);
  if (termMustNotQuery) {
    mustNot.push(termMustNotQuery);
  }
  Object.keys(fields.mustNot.and).forEach(field => {
    mustNot.push(fieldValuesToQuery(field, fields.mustNot.and[field], 'and'));
  });
  Object.keys(fields.mustNot.or).forEach(field => {
    mustNot.push(fieldValuesToQuery(field, fields.mustNot.or[field], 'or'));
  });

  const bool = {};
  if (must.length !== 0) {
    bool.must = must;
  }
  if (mustNot.length !== 0) {
    bool.must_not = mustNot;
  }
  return { bool };
};
