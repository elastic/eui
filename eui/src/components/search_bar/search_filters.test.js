import React from 'react';
import { requiredProps } from '../../test';
import { shallow } from 'enzyme';
import { EuiSearchFilters } from './search_filters';
import { Query } from './query';

describe('EuiSearchFilters', () => {

  test('render - no filters', () => {

    const props = {
      ...requiredProps,
      onChange: () => {},
      query: Query.parse(''),
      filters: []
    };

    const component = shallow(
      <EuiSearchFilters {...props} />
    );

    expect(component).toMatchSnapshot();

  });

  test('render - with filters', () => {

    const props = {
      ...requiredProps,
      onChange: () => {},
      query: Query.parse(''),
      filters: [
        {
          type: 'is',
          field: 'open',
          name: 'Open'
        },
        {
          type: 'field_value_selection',
          field: 'tag',
          name: 'Tag',
          options: () => {}
        }
      ]
    };

    const component = shallow(
      <EuiSearchFilters {...props} />
    );

    expect(component).toMatchSnapshot();
  });

});
