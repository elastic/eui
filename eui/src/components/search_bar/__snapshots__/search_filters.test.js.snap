// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSearchFilters render - no filters 1`] = `<EuiFilterGroup />`;

exports[`EuiSearchFilters render - with filters 1`] = `
<EuiFilterGroup>
  <React.Fragment
    key="filter_0"
  >
    <IsFilter
      config={
        Object {
          "field": "open",
          "name": "Open",
          "type": "is",
        }
      }
      index={0}
      onChange={[Function]}
      query={
        Query {
          "ast": _AST {
            "_clauses": Array [],
            "_indexedClauses": Object {
              "field": Object {},
              "is": Object {},
              "term": Array [],
            },
          },
          "syntax": Object {
            "parse": [Function],
            "print": [Function],
          },
          "text": "",
        }
      }
    />
  </React.Fragment>
  <React.Fragment
    key="filter_1"
  >
    <FieldValueSelectionFilter
      autoClose={true}
      config={
        Object {
          "field": "tag",
          "name": "Tag",
          "options": [Function],
          "type": "field_value_selection",
        }
      }
      index={1}
      onChange={[Function]}
      query={
        Query {
          "ast": _AST {
            "_clauses": Array [],
            "_indexedClauses": Object {
              "field": Object {},
              "is": Object {},
              "term": Array [],
            },
          },
          "syntax": Object {
            "parse": [Function],
            "print": [Function],
          },
          "text": "",
        }
      }
    />
  </React.Fragment>
</EuiFilterGroup>
`;
